modules/data-streams/src/main/java/org/elasticsearch/datastreams/lifecycle/DataStreamLifecycleErrorStore.java

```diff
@@ -48,21 +49,22 @@ public DataStreamLifecycleErrorStore(LongSupplier nowSupplier) {
      * Returns the previously recorded error for the provided index, or null otherwise.
      */
     @Nullable
-    public ErrorEntry recordError(String indexName, Exception e) {
+    public ErrorEntry recordError(ProjectId projectId, String indexName, Exception e) {
         String exceptionToString = Strings.toString((builder, params) -> {
             ElasticsearchException.generateThrowableXContent(builder, EMPTY_PARAMS, e);
             return builder;
         });
         String newError = Strings.substring(exceptionToString, 0, MAX_ERROR_MESSAGE_LENGTH);
-        ErrorEntry existingError = indexNameToError.get(indexName);
+        final var projectMap = indexNameToError.computeIfAbsent(projectId, k -> new ConcurrentHashMap<>());
+        ErrorEntry existingError = projectMap.get(indexName);
```

Comment:
Nit: I personally feel the variable names `projectMap` and `indexNameToError` should be swapped.